### PR TITLE
*: add engine filter to exclude special stores

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve? 
Fix https://github.com/pingcap/pd/issues/2423
`AvaiableStore()` always returns flash nodes, so the tikv nodes cannot be used.

### Release note 
- Fix the issue that `presplit` not work within tiflash cluster.

### CODE REVIEW

File: server/schedule/filter/filters.go [hunk 1]
```diff
@@ -543,6 +543,41 @@ func (f *ruleFitFilter) Target(opt opt.Options, store *core.StoreInfo) bool {
 	return placement.CompareRegionFit(f.oldFit, newFit) <= 0
 }
 
+type engineFilter struct {
+	scope      string
+	constraint placement.LabelConstraint
+}
+
+// NewEngineFilter creates a filter that filters store for special engine.
+func NewEngineFilter(scope string, allowUses ...string) Filter {
```
Comment (HunDunDM) [inline server/schedule/filter/filters.go:552]: Can constraint.Values be specified directly in the constructor?

Comment (nolouch) [reply server/schedule/filter/filters.go]: I have adds some comments.

File: server/schedule/filter/filters.go [hunk 2]
```diff
@@ -543,12 +543,51 @@ func (f *ruleFitFilter) Target(opt opt.Options, store *core.StoreInfo) bool {
 	return placement.CompareRegionFit(f.oldFit, newFit) <= 0
 }
 
+type engineFilter struct {
+	scope      string
+	constraint placement.LabelConstraint
+}
+
+// NewEngineFilter creates a filter that filters out default engine stores.
+// By default, all stores that are not marked with a special engine will be filtered out.
```
Comment (lhy1024) [inline server/schedule/filter/filters.go:552]: all tikv node will be filtered out?

File: server/schedule/filter/filters.go [hunk 3]
```diff
@@ -543,12 +543,51 @@ func (f *ruleFitFilter) Target(opt opt.Options, store *core.StoreInfo) bool {
...
+// Specify the special engine label if you want to include the special stores.
+func NewEngineFilter(scope string, allowSpeicalEngines ...string) Filter {
```
Comment (rleungx) [inline server/schedule/filter/filters.go:554]: There are too many inverse logic and it's hard to understand.

File: server/schedule/filter/filters.go (hunk 2 above)
Comment (nolouch) [reply server/schedule/filter/filters.go:552]: yes.

File: server/schedule/filter/filters.go (hunk 3 above)
Comment (nolouch) [reply server/schedule/filter/filters.go:554]: any suggestion?

**Comment by nolouch** [comment]
It's about dashboard, merge it firstly.

**Pull request merged**